services/api/src/server.ts

```
@@ -16,17 +16,145 @@ import { keycloakHasPermission, legacyHasPermission } from './util/auth';
+        logger.error('SSE: No authorization header');
+        throw new AuthenticationError('Auth token missing.');
+      }
+
+      const authToken = authHeader.replace('Bearer ', '').replace('bearer ', '');
+
+      try {
+        const { grant, legacyCredentials } = await getGrantOrLegacyCredsFromToken(authToken);
+        return { grant, legacyCredentials };
+      } catch (e) {
+        logger.error(`SSE auth failed: ${e.message}`);
+        throw new AuthenticationError('Auth token invalid.');
+      }
+    },
+    context: async (req, params) => {
+      const { grant: keycloakGrant, legacyCredentials: legacyGrant } = req.context || {};
+
+      const keycloakAdminClient = await getKeycloakAdminClient();
+      const modelClients = { sqlClientPool, keycloakAdminClient, esClient };
+
+      let currentUser: any = { id: undefined };
+      let serviceAccount = {};
+      let keycloakUsersGroups = [];
+      let groupRoleProjectIds = [];
+      let platformOwner = false;
+      let platformViewer = false;
+
+      if (keycloakGrant) {
+        try {
+          const userId = keycloakGrant.access_token.content.sub;
+          keycloakUsersGroups = await User(modelClients).getAllGroupsForUser(userId);
+          serviceAccount = await keycloakGrantManager.obtainFromClientCredentials(undefined, undefined);
+          currentUser = await User(modelClients).loadUserById(userId);
+          const userRoleMapping = await keycloakAdminClient.users.listRealmRoleMappings({ id: currentUser.id });
+          platformOwner = userRoleMapping.some(role => role.name === 'platform-owner');
+          platformViewer = userRoleMapping.some(role => role.name === 'platform-viewer');
+          groupRoleProjectIds = await User(modelClients).getAllProjectsIdsForUser(currentUser.id, keycloakUsersGroups);
+          await User(modelClients).userLastAccessed(currentUser);
+        } catch (e) {
+          logger.error('Error loading user details for SSE subscription', e.message);
```

We're just logging an error here and still creating the `context`, can we throw a new error if we can't get the user details?

---

We do the same thing in the existing check [here](https://github.com/uselagoon/lagoon/blob/v2.29.2/services/api/src/server.ts#L83) which is unmodified.

If it is changed here, it probably needs to also be changed in that section too. Do you have a suggestion on how that may work?

---

I was thinking of simply throwing an error? 

```
  } catch (e) {
    logger.error('Error loading user details for SSE subscription', e.message);
    throw new AuthenticationError('Failed to load user context for SSE subscription');
  }
```

We're doing the same in the [onConnect](https://github.com/uselagoon/lagoon/blob/main/services/api/src/server.ts#L150) for the `websocket` setup & the [authenticate](https://github.com/uselagoon/lagoon/pull/4007/files#diff-a0949ca6ce80ca47adf27b3eb9d29920fa54cc25273251f72e3ea133afa416e3R47) for the `sseHandler`, can probably do the same for the `websocket context`. Not sure if there's further implications for doing this though?

---

Probably fine to throw an error, if the user isn't authenticated, it'll fail somewhere further down anyway.